src/fabric/src/fabric2_fdb.erl

```diff
@@ -647,19 +652,29 @@ write_doc(#{} = Db0, Doc, NewWinner0, OldWinner, ToUpdate, ToRemove) ->
 
 write_local_doc(#{} = Db0, Doc) ->
     #{
-        tx := Tx
+        tx := Tx,
+        db_prefix := DbPrefix
     } = Db = ensure_current(Db0),
 
-    {LDocKey, LDocVal} = local_doc_to_fdb(Db, Doc),
+    Id = Doc#doc.id,
+
+    {LDocKey, LDocVal, Rows} = local_doc_to_fdb(Db, Doc),
+
+    upgrade_local_doc(Db, Id),
 
     WasDeleted = case erlfdb:wait(erlfdb:get(Tx, LDocKey)) of
         <<_/binary>> -> false;
         not_found -> true
     end,
 
     case Doc#doc.deleted of
-        true -> erlfdb:clear(Tx, LDocKey);
-        false -> erlfdb:set(Tx, LDocKey, LDocVal)
+        true ->
+            erlfdb:clear(Tx, LDocKey),
+            Prefix = erlfdb_tuple:pack({?DB_LOCAL_DOC_BODIES, Id}, DbPrefix),
+            erlfdb:clear_range_startswith(Tx, Prefix);
+        false ->
+            erlfdb:set(Tx, LDocKey, LDocVal),
+            lists:foreach(fun({K, V}) -> erlfdb:set(Tx, K, V) end, Rows)
```

Comment:
You should add a test for a large local doc that is then shrunk to fewer chunks. Currently you're not clearing the extra chunks from the previous version so you'd likely get a badarg in `binary_to_term` down below.

Reply:
Ooh. Good catch. And good idea to add a test for it

Reply:
Added a test:

https://github.com/apache/couchdb/pull/2267/files#diff-8488ec220e8295563d8258f4bc2198adR797